README.md
# optics-agent-ruby
Apollo Optics agent for GraphQL Monitoring in Ruby.

[![Gem Version](https://badge.fury.io/rb/optics-agent.svg)](https://badge.fury.io/rb/optics-agent) [![Build Status](https://travis-ci.org/apollostack/optics-agent-ruby.svg?branch=master)](https://travis-ci.org/apollostack/optics-agent-ruby)


## Installing

Add

```ruby
gem 'optics-agent'
```

To your `Gemfile`

### API key

You'll need to run your app with the `OPTICS_API_KEY` environment variable set (or set via [`agent.configure`](#configuration)) to the API key of your Apollo Optics endpoint; you can get an API key by setting up a endpoint at https://optics.apollodata.com.

## Rails Setup

Create an agent in `config/initializers/optics_agent.rb`, and register the rack middleware:
```ruby
optics_agent = OpticsAgent::Agent.new
optics_agent.configure do
  schema YourSchema
  # See other configuration options below
end
Rails.application.config.middleware.use optics_agent.rack_middleware
```

Register Apollo Optics agent from your on the GraphQL context within your `graphql` action as below:
```ruby
def create
  query_string = params[:query]
  query_variables = ensure_hash(params[:variables])

  result = YourSchema.execute(
    query_string,
    variables: query_variables,
    context: {
      # This is the key line: we take the optics_agent passed in from the
      # Rack environment and pass it as context
      optics_agent: env[:optics_agent]
    }
  )

  render json: result
end
```

You can check out the GitHunt Rails API server example here: https://github.com/apollostack/githunt-api-rails

## General Setup

You must:

1. Create an agent with `OpticsAgent::Agent.new`
2. Register your schema with the `agent.configure` block
3. Attach the `agent.rack_middleware` to your Rack router
4. Ensure you pass the `optics_agent` context from the rack environment to your schema execution.

### Non-HTTP queries

If you aren't actually servicing a HTTP/Rack request in executing the query, simply pass:

```ruby
  context: { optics_agent: :no_rack }
```

This will mean the query is instrumented without HTTP timings or client versions.

### Skipping queries

If you want to skip a particular query, pass:

```ruby
  context: { optics_agent: :skip }
```

### Configuration

After creating an agent, you can configure it with (the values listed are the defaults):

```rb
# defaults are show below
agent.configure do
  # The schema you wish to instrument
  schema YourSchema

  # Your API key for the Optics service. This defaults to the OPTICS_API_KEY
  # environment variable, but can be overridden here.
  api_key ENV['OPTICS_API_KEY']

  # Log detailed debugging messages
  debug false

  # Don't report anything to Optics (useful for testing)
  disable_reporting false

  # Print JSON versions of the data sent to Optics to the log
  print_reports false

  # Send detailed traces along with usage reports
  report_traces true

  # How long to wait before sending a schema report after startup, in
  # milliseconds
  schema_report_delay_ms 10 * 1000

  # How often to send reports in milliseconds. Defaults to 1 minute.
  # You shouldn't need to set this unless you are debugging.
  report_interval_ms 60 * 1000

  # Where to send the reports. Defaults to the production Optics endpoint,
  # or the `OPTICS_ENDPOINT_URL` environment variable if it is set.
  # You shouldn't need to set this unless you are debugging
  endpoint_url 'https://optics-report.apollodata.com'
end
```

## Troubleshooting

The Apollo Optics agent is designed to allow your application to continue working, even if the agent is not configured properly.

### No data in Optics

If there is no data being sent to Optics, ensure you've followed the steps above, then check your application logs to look for the following messages:

### Message: No api_key set.

Solution: Get a valid API key from Optics and use the `OPTICS_API_KEY` environment variable, or [configuration](#configuration) to pass it to the agent.

### Message: No schema instrumented

Solution: Ensure you are passing your schema to the agent [configuration](#configuration).

### Message: No agent passed in graphql context

Solution: Ensure you are passing `context: { optics_agent: env[:optics_agent].with_document(query_string) }` where `env` is the Rack request environment, and `query_string` is the string representing your query. See the [setup instructions](#rails-setup) for more details.

### Debugging

You can also use the `debug` [configuration](#configuration) setting to get more detailed debugging information, which may give hints as to what the issue is.

## Development

### Running tests

```
bundle install
bundle exec rspec
```

### Building protobuf definitions

Ensure you've installed `protobuf` and development dependencies.

```bash
# this works on OSX
brew install protobuf

# ensure it's version 3
protoc --version

bundle install
````

Compile the `.proto` definitions with

```bash
bundle exec rake protobuf:compile
```


CHANGELOG.md
### vNEXT

### v0.5.4

- Bugfixes.

### v0.5.3

- Support for non-HTTP queries
- More resilient communication with the Optics service

### v0.4.4

- Fix issue with `graphql@1.2.5` due to breaking change in visitor API. (Issue [#42](https://github.com/apollostack/optics-agent-ruby/issues/42))

### v0.4.3

- Work around issue with graphl-ruby instrumenters and Relay connection fields. (See [#417](https://github.com/rmosolgo/graphql-ruby/issues/417)).

### v0.4.2

- Loosened unnecessarily strict dependencies in gemspec


### v0.4.1
- Fix issue where queries did not execute if you mis-configured the agent (now it just warns).

- Use the Rails logger if available and allow setting `OpticsAgent::Agent.logger`.

### v0.4.0

#### Breaking

- Requires `graphl-ruby@1.1.0`

- We have a new configuration setup, read it [here](https://github.com/apollostack/optics-agent-ruby#rails-setup)

  - Use `OpticsAgent::Agent.new` rather than `OpticsAgent::Agent.instance`

  - Use `agent.configure`, with a [DSL](https://github.com/apollostack/optics-agent-ruby#configuration) rather than `agent.set_options`

#### New features

- Sample the traces we report [Issue #20](https://github.com/apollostack/optics-agent-ruby/issues/20), rather than reporting a trace for every query.

- Use `hitimes`, a high resolution timer library [Issue #21](https://github.com/apollostack/optics-agent-ruby/issues/21).

- You can now disable trace reporting with the `report_traces` option

### v0.3.1

- Be more aggressive about waiting until the first request to spawn the reporting thread; to achieve better behaviour on pre-forking webservers such as Unicorn and Puma.


### v0.3.0

- Added `agent.set_options` to allow you to configure the agent. (See the README for more info)
- Added a `debug` option which logs a lot of useful info about reporting
- Use an object for :optics_agent [PR #28](https://github.com/apollostack/optics-agent-ruby/pull/28)

### v0.2.1

- Use a simple `Thread.new` instead of sucker_punch and concurrent-ruby.

### v0.2.0

- We only require you to pass in the query_string, rather than the full graphql request params now. See https://github.com/apollostack/optics-agent-ruby/pull/27

### v0.1.3

- Wait until the schema is added to start the reporting loop. This means of pre-forking webservers like unicorn, we'll wait until the process has forked and started a "true" process.


lib/optics-agent/agent.rb
require 'optics-agent/rack-middleware'
require 'optics-agent/instrumenters/field'
require 'optics-agent/reporting/report_job'
require 'optics-agent/reporting/schema_job'
require 'optics-agent/reporting/query-trace'
require 'optics-agent/reporting/detect_server_side_error'
require 'net/http'
require 'faraday'
require 'logger'
require 'time'

module OpticsAgent
  class Agent
    include OpticsAgent::Reporting

    class << self
      attr_accessor :logger
    end

    self.logger = Logger.new(STDOUT)

    attr_reader :schema, :report_traces

    def initialize
      @stats_reporting_thread = nil
      @query_queue = []
      @semaphone = Mutex.new

      # set defaults
      @configuration = Configuration.new
    end

    def configure(&block)
      @configuration.instance_eval(&block)

      if @configuration.schema && @schema != @configuration.schema
        instrument_schema(@configuration.schema)
      end
    end

    def disabled?
      @configuration.disable_reporting || !@configuration.api_key || !@schema
    end

    def report_traces?
      @configuration.report_traces
    end

    def instrument_schema(schema)
      unless @configuration.api_key
        warn """No api_key set.
Either configure it or use the OPTICS_API_KEY environment variable.
"""
        return
      end

      if @schema
        warn """Agent has already instrumented a schema.
Perhaps you are calling both `agent.configure { schema YourSchema }` and
`agent.instrument_schema YourSchema`?
"""
        return
      end

      @schema = schema
      @schema._attach_optics_agent(self)

      unless disabled?
        debug "spawning schema thread"
        Thread.new do
          debug "schema thread spawned"
          sleep @configuration.schema_report_delay_ms / 1000.0
          debug "running schema job"
          SchemaJob.new.perform(self)
        end
      end
    end

    # We call this method on every request to ensure that the reporting thread
    # is active in the correct process for pre-forking webservers like unicorn
    def ensure_reporting_stats!
      unless @schema
        warn """No schema instrumented.
Use the `schema` configuration setting, or call `agent.instrument_schema`
"""
        return
      end

      unless reporting_stats? || disabled?
        schedule_report
      end
    end

    # @deprecated Use ensure_reporting_stats! instead
    def ensure_reporting!
      ensure_reporting_stats!
    end

    def reporting_connection
      @reporting_connection ||=
        Faraday.new(:url => @configuration.endpoint_url) do |conn|
          conn.request :retry,
                       :max => 5,
                       :interval => 0.1,
                       :max_interval => 10,
                       :backoff_factor => 2,
                       :exceptions => [Exception],
                       :retry_if => ->(env, exc) { true }
          conn.use OpticsAgent::Reporting::DetectServerSideError

          # XXX: allow setting adaptor in config
          conn.adapter :net_http_persistent
        end
    end

    def reporting_stats?
      @stats_reporting_thread != nil && !!@stats_reporting_thread.status
    end

    def schedule_report
      @semaphone.synchronize do
        return if reporting_stats?

        debug "spawning reporting thread"

        thread = Thread.new do
          begin
            debug "reporting thread spawned"

            report_interval = @configuration.report_interval_ms / 1000.0
            last_started = Time.now

            while true
              next_send = last_started + report_interval
              sleep_time = next_send - Time.now

              if sleep_time < 0
                warn 'Report took more than one interval! Some requests might appear at the wrong time.'
              else
                sleep sleep_time
              end

              debug "running reporting job"
              last_started = Time.now
              ReportJob.new.perform(self)
              debug "finished running reporting job"
            end
          rescue Exception => e
            warn "Stats report thread dying: #{e}"
          end
        end

        at_exit do
          if thread.status
            debug 'sending last stats report before exiting'
            thread.exit
            ReportJob.new.perform(self)
            debug 'last stats report sent'
          end
        end

        @stats_reporting_thread = thread
      end
    end

    def add_query(*args)
      @semaphone.synchronize {
        debug { "adding query to queue, queue length was #{@query_queue.length}" }
        @query_queue << args
      }
    end

    def clear_query_queue
      @semaphone.synchronize {
        debug { "clearing query queue, queue length was #{@query_queue.length}" }
        queue = @query_queue
        @query_queue = []
        queue
      }
    end

    def rack_middleware
      # We need to pass ourselves to the class we return here because
      # rack will instanciate it. (See comment at the top of RackMiddleware)
      OpticsAgent::RackMiddleware.agent = self
      OpticsAgent::RackMiddleware
    end

    def graphql_middleware
      warn "You no longer need to pass the optics agent middleware, it now attaches itself"
    end

    def send_message(path, message)
      response = reporting_connection.post do |request|
        request.url path
        request.headers['x-api-key'] = @configuration.api_key
        request.headers['user-agent'] = "optics-agent-rb"
        request.body = message.class.encode(message)

        if @configuration.debug || @configuration.print_reports
          log "sending message: #{message.class.encode_json(message)}"
        end
      end

      if @configuration.debug || @configuration.print_reports
        log "got response body: #{response.body}"
      end
    end

    def log(message = nil)
      message = yield unless message
      self.class.logger.info "optics-agent: #{message}"
    end

    def warn(message = nil)
      self.class.logger.warn "optics-agent: WARNING: #{message}"
    end

    # Should we be using built in debug levels rather than our own "debug" flag?
    def debug(message = nil)
      if @configuration.debug
        message = yield unless message
        self.class.logger.info "optics-agent: DEBUG: #{message} <#{Process.pid} | #{Thread.current.object_id}>"
      end
    end
  end

  class Configuration
    def self.defaults
      {
        schema: nil,
        debug: false,
        disable_reporting: false,
        print_reports: false,
        report_traces: true,
        schema_report_delay_ms: 10 * 1000,
        report_interval_ms: 60 * 1000,
        api_key: ENV['OPTICS_API_KEY'],
        endpoint_url: ENV['OPTICS_ENDPOINT_URL'] || 'https://optics-report.apollodata.com'
      }
    end

    # Allow e.g. `debug false` == `debug = false` in configuration blocks
    defaults.each_key do |key|
      define_method key, ->(*maybe_value) do
        if (maybe_value.length === 1)
          self.instance_variable_set("@#{key}", maybe_value[0])
        elsif (maybe_value.length === 0)
          self.instance_variable_get("@#{key}")
        else
          throw new ArgumentError("0 or 1 argument expected")
        end
      end
    end

    def initialize
      self.class.defaults.each { |key, value| self.send(key, value) }
    end
  end
end
